common/include/common/refcount.h

```diff
@@ -0,0 +1,30 @@
+/*
+Looking Glass - KVM FrameRelay (KVMFR) Client
+Copyright (C) 2021 Guanzhong Chen <quantum2048@gmail.com>
+https://looking-glass.hostfission.com
+
+This program is free software; you can redistribute it and/or modify it under
+the terms of the GNU General Public License as published by the Free Software
+Foundation; either version 2 of the License, or (at your option) any later
+version.
+
+This program is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+Place, Suite 330, Boston, MA 02111-1307 USA
+*/
+
+#include <stddef.h>
+
+struct CountedBuffer {
+  _Atomic(size_t) refs;
+  size_t size;
+  char * data;
+};
+
+struct CountedBuffer * countedBufferCreate(size_t size);
```

Comment:
Please use the following prototype for this instead, allowing the caller to provide the memory for `struct CountedBuffer` to avoid the need to double malloc. Failure should zero `return`.

    bool countedBufferNew(size_t size, struct CountedBuffer * return)

Reply:
Renamed to `countedBufferNew`.

Having the caller memory the memory for the `struct CountedBuffer` defeats the purpose of the entire thing, which is to allow the caller to call release and throw its reference away, while keeping the buffer around if something else wants it.

If you meant to have the caller `malloc` the structure, then have `countedBufferRelease` `free` the structure, I still think it's awkward, since the memory management is no longer handled in a single module.

Reply:
This is to allow a static buffer to be used/reused, the actual data member is still allocated via `malloc` by `counterBufferNew`, for example:

```
static struct CountedBuffer buffer;

int main()
{
  if (!countedBufferNew(1024, &buffer))
    return 1;
  return 0;
}
```

Reply:
This use case would not work. The intended use case is to keep past clipboard content around for ongoing transfers while the current reference is updated to the latest content. Statically allocating the `struct CountedBuffer` prevents us from such a use case, since we need multiple `struct CountedBuffer` with overlapping lifetimes.

Reply:
Sorry yes, I follow its usage pattern now, I am just not keen on the double malloc. What about keeping the data contained within the struct?, ie:

```
struct CountedBuffer
{
struct CountedBuffer {
  _Atomic(size_t) refs;
  size_t size;
  // perhaps pad to ensure alignment, however this should be aligned anyway by gcc
  char data[];
};

data = malloc(sizeof(struct CountedBuffer) + size);
```

Reply:
Yea this sounds good. Changed to use the flexible array member.